mergekit/scripts/extract_lora.py

```diff
@@ -0,0 +1,194 @@
+        "inference_mode": True,
+        "init_lora_weights": True,
+        "layers_pattern": None,
+        "layers_to_transform": None,
+        "loftq_config": {},
+        "lora_alpha": alpha,
+        "lora_dropout": 0,
+        "megatron_config": None,
+        "megatron_core": "megatron.core",
+        "modules_to_save": None,
+        "peft_type": "LORA",
+        "r": rank,
+        "rank_pattern": {},
+        "revision": None,
+        "target_modules": target_modules,
+        "task_type": "CAUSAL_LM",
+        "use_rslora": False,
+    }
+
+
+@click.command("mergekit-extract-lora")
+@click.argument("out_path", type=click.Path())
+@click.option("--base-model", type=str, help="Model ID or path to use as base model")
+@click.option(
+    "--finetuned-model",
+    type=str,
+    help="Model ID or path to use as PEFT extraction target model",
+)
+@click.option(
+    "--rank", type=int, default=32, help="Rank for the low-rank decomposition"
+)
+@click.option(
+    "--device",
+    type=str,
+    default=None,
+    help="PyTorch device to perform SVD computation on",
+)
+def main(out_path: str, base_model: str, finetuned_model: str, rank: int, device: str):
+    """
+    Decomposes delta weights between a base model and a finetuned model and saves a PEFT model.
+    """
+    os.makedirs(out_path, exist_ok=True)
+
+    base_model_ref = ModelReference.parse(base_model)
+    finetuned_model_ref = ModelReference.parse(finetuned_model)
+
+    base_model_config = AutoConfig.from_pretrained(base_model_ref.model.path)
+
+    linear_module_names = get_linear_module_names(base_model_ref.model.path)
+    finetuned_model_linear_module_names = get_linear_module_names(
+        finetuned_model_ref.model.path
+    )
+
+    assert set(linear_module_names) == set(
+        finetuned_model_linear_module_names
+    ), "Model architecture mismatch"
+
+    base_loader = LazyTensorLoader(base_model_ref.tensor_index(), lazy_unpickle=True)
+    finetuned_loader = LazyTensorLoader(
+        finetuned_model_ref.tensor_index(), lazy_unpickle=True
+    )
+
+    lora_weights = {}
+    for layer_name in tqdm(linear_module_names):
+        base_weight = base_loader.get_tensor(f"{layer_name}.weight")
+        finetuned_weight = finetuned_loader.get_tensor(f"{layer_name}.weight")
+
+        lora_A, lora_B = decompose_delta_weight(
+            finetuned_weight, base_weight, rank, device=device
+        )
+
+        lora_weights[f"base_model.model.{layer_name}.lora_A.weight"] = lora_A.to(
+            "cpu"
+        ).contiguous()
+        lora_weights[f"base_model.model.{layer_name}.lora_B.weight"] = lora_B.to(
+            "cpu"
+        ).contiguous()
+
+    lora_config = create_peft_config(
+        base_model_name_or_path=base_model_ref.model.path,
+        alpha=rank,  # Setting the alpha to the reduced rank value (instead of alpha value used) seems to give better performance. Further testing would be needed to understand what is the optimal alpha value to use
```

Comment:
This is correct yeah! The delta is scaled by `alpha/r` (see [here](https://github.com/huggingface/peft/blob/02b5aeddf9c1ea11451f10a8a26da7e5df8cca4a/src/peft/tuners/lora/layer.py#L107)) so `alpha=r` is going to give you the closest possible approximation of the fine tuned model when applied to the base.